### PR TITLE
Add forwardRef2/memo2 which fixes jsification of Dart props

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
# Motivation
Dart `forwardRef` components and `memo`-wrapped components were being passed JSified props due to the use of JS component factories, when they expected to be getting non-converted Dart props.

forwardRef-memo-bug

# Solution
__tl;dr: I added `forwardRef2`/`memo2`, added a bunch of tests around refs, and fixed a bunch of issues surfaced by test failures.__

- Add `forwardRef2` and `memo2`, which return Dart component factories so they don't convert props
    - I had to make new versions because fixing the issues required return type changes that would be breaking. We can consume these new versions under the hood in over_react's memo/uiForwardRef without any breakages.
    - Remove prop de-converting logic from `forwardRef2`
        This was only a partial solution to wrapping JS components, only applying to callbacks

        Also, now that we're using Dart component factories, this logic would prevent Dart event handlers from being converted.

        So, since we'll need a more holistic solution to the problem of wrapping JS components, and it wouldn't be easy to retain this logic without breaking other behavior, it was removed.

- Update common factory tests:
    - Refactor simple ref tests use ref test case objects for consistency and to DRY up setup
    - Add tests for all Dart factories (including forwardRef2/memo2) that Dart props are passed in without being converted
        These verify that the main issue that motivated this PR (converted JS props being passed to Dart components) is fixed
    - Add tests that verify that all ref types work with forwardRef
    - Add tests that `chainRef` works with refs that come from forwardRef2 callback args (as well as ReactElement, since it experienced a similar problem)
        - This required some code changes in different places to get these passing
    - Add test cases JS ref objects are populated with the correct values

- Update event handler tests to not test de-conversion of props via `forwardRef`

# QA Instructions
- Verify unit tests pass (CI passes)
- Verify the function components in WSD that originally surfaced these issues still work (@aaronlademann-wf)
- Smoke test examples
- Perform QA instructions in over_react PR consuming these in `uiForwardRef` and `memo`: https://github.com/Workiva/over_react/pull/620

### CODE REVIEW

File: lib/react_client/component_factory.dart [hunk 1]
```diff
@@ -367,34 +368,83 @@ class ReactDartFunctionComponentFactoryProxy extends ReactComponentFactoryProxy
 
   @override
   JsFunctionComponent get type => reactFunction;
+}
 
-  static String _getJsFunctionName(Function object) =>
```
Comment (greglittlefield-wf) [inline lib/react_client/component_factory.dart:371]: Use [whitespace-agnostic diff](https://github.com/cleandart/react-dart/pull/275/files?w=1) for this section

File: test/lifecycle_test.dart [hunk 2]
```diff
@@ -551,24 +551,6 @@ main() {
           expect(propTypeCheck, isA<JsBackedMap>());
         });
       });
-
-      group('props are received correctly without interop interfering with the values:', () {
```
Comment (greglittlefield-wf) [inline test/lifecycle_test.dart:555]: These were moved to the common factory tests so they run for all Dart components and Dart component HOCs, not just Dart function components

File: lib/react_client/component_factory.dart [hunk 3]
```diff
@@ -172,10 +159,14 @@ class ReactDartComponentFactoryProxy<TComponent extends Component> extends React
...
-        interopProps.ref = allowInterop((ReactComponent instance) {
+        interopProps.ref = allowInterop((dynamic instance) {
           // Call as dynamic to perform dynamic dispatch, since we can't cast to CallbackRef<dynamic>,
           // and since calling with non-null values will fail at runtime due to the CallbackRef<Null> typing.
-          return (ref as dynamic)(instance?.dartComponent);
+          if (instance is ReactComponent && instance.dartComponent != null) {
+            return (ref as dynamic)(instance.dartComponent);
+          }
+
+          return (ref as dynamic)(instance);
```
Comment (greglittlefield-wf) [inline lib/react_client/component_factory.dart:169]: These changes update ReactDartComponentFactoryProxy's ref wrapping code to that of ReactDartComponentFactoryProxy**2**, and were made to fix `chainRef` tests

File: lib/react_client/component_factory.dart [hunk 4]
```diff
@@ -36,20 +37,6 @@ dynamic listifyChildren(dynamic children) {
...
 
-/// Returns the original Dart handler function that, within [_convertEventHandlers],
-/// was converted/wrapped into the function [jsConvertedEventHandler] to be passed to the JS.
-///
-/// Returns `null` if [jsConvertedEventHandler] is `null`.
-///
-/// Returns `null` if [jsConvertedEventHandler] does not represent such a function
-///
-/// Useful for chaining event handlers on DOM or JS composite [ReactElement]s.
-Function unconvertJsEventHandler(Function jsConvertedEventHandler) {
```
Comment (greglittlefield-wf) [inline lib/react_client/component_factory.dart:47]: I moved this into factory_util.dart so that it would live next to event conversion logic.

I did this when I had similar unconversion logic for refs, but ended up not needing that new logic. Instead of moving `unconvertJsEventHandler` back, I thought I'd keep it in its new home.

**Comment by semveraudit-wf** [comment]
### Public API Changes

 Recommendation: **:bangbang: Major version bump (fyi @Workiva/semver-audit-group )**


```diff
@@ line 37: package:react/react_client/react_interop.dart @@
abstract class React
-     static ReactClass memo(dynamic Function(JsMap, [JsMap]) wrapperFunction, [bool Function(JsMap, JsMap) areEqual])
+     static ReactClass memo(dynamic wrapperFunction, [bool Function(JsMap, JsMap) areEqual])
//    `type` of `wrapperFunction` has changed.
//    Changing a parameter signature is a major change.

```

```diff
@@ line 38: package:react/react.dart @@
+  typedef DartForwardRefFunctionComponent = dynamic Function(JsBackedMap props, dynamic ref)
// Adding a typedef is a minor change.
```

```diff
@@ line 232: package:react/react_client/react_interop.dart @@
+  ReactComponentFactoryProxy forwardRef2(dynamic Function(Map<dynamic, dynamic>, dynamic) wrapperFunction, {String displayName})
// Adding a top-level function is a minor change.
```

```diff
@@ line 232: package:react/react_client/react_interop.dart @@
+  ReactComponentFactoryProxy forwardRef2(dynamic Function(Map<dynamic, dynamic>, dynamic) wrapperFunction, {String displayName})
// Adding a top-level function is a minor change.
```

```diff
@@ line 303: package:react/react_client/react_interop.dart @@
+  ReactComponentFactoryProxy memo2(ReactComponentFactoryProxy factory, {bool Function(Map<dynamic, dynamic>, Map<dynamic, dynamic>) areEqual})
// Adding a top-level function is a minor change.
```
<details>
<summary>Click to see 2 more API Changes</summary><p>

---
```diff
@@ line 303: package:react/react_client/react_interop.dart @@
+  ReactComponentFactoryProxy memo2(ReactComponentFactoryProxy factory, {bool Function(Map<dynamic, dynamic>, Map<dynamic, dynamic>) areEqual})
// Adding a top-level function is a minor change.
```

```diff
@@ line 364: package:react/react_client/component_factory.dart @@
+  class ReactDartWrappedComponentFactoryProxy extends ReactComponentFactoryProxy with JsBackedMapComponentFactoryMixin
// Adding a class is a minor change.

+     ReactClass get type
//    Adding a field is a minor change.

+     ReactElement build(Map<dynamic, dynamic> props, [List<dynamic> childrenArgs = const []])
//    Adding a method is a minor change.

+     ReactDartWrappedComponentFactoryProxy ReactDartWrappedComponentFactoryProxy.forwardRef(dynamic Function(JsBackedMap, dynamic) dartFunctionComponent, {String displayName})
//    Adding a constructor is a minor change.

+     ReactDartWrappedComponentFactoryProxy ReactDartWrappedComponentFactoryProxy(ReactClass type)
//    Adding a constructor is a minor change.

      From package:react/react.dart
+     dynamic call(Map<dynamic, dynamic> props, [dynamic c1 = _notSpecified, dynamic c2 = _notSpecified, dynamic c3 = _notSpecified, dynamic c4 = _notSpecified, dynamic c5 = _notSpecified, dynamic c6 = _notSpecified, dynamic c7 = _notSpecified, dynamic c8 = _notSpecified, dynamic c9 = _notSpecified, dynamic c10 = _notSpecified, dynamic c11 = _notSpecified, dynamic c12 = _notSpecified, dynamic c13 = _notSpecified, dynamic c14 = _notSpecified, dynamic c15 = _notSpecified, dynamic c16 = _notSpecified, dynamic c17 = _notSpecified, dynamic c18 = _notSpecified, dynamic c19 = _notSpecified, dynamic c20 = _notSpecified, dynamic c21 = _notSpecified, dynamic c22 = _notSpecified, dynamic c23 = _notSpecified, dynamic c24 = _notSpecified, dynamic c25 = _notSpecified, dynamic c26 = _notSpecified, dynamic c27 = _notSpecified, dynamic c28 = _notSpecified, dynamic c29 = _notSpecified, dynamic c30 = _notSpecified, dynamic c31 = _notSpecified, dynamic c32 = _notSpecified, dynamic c33 = _notSpecified, dynamic c34 = _notSpecified, dynamic c35 = _notSpecified, dynamic c36 = _notSpecified, dynamic c37 = _notSpecified, dynamic c38 = _notSpecified, dynamic c39 = _notSpecified, dynamic c40 = _notSpecified])
//    Adding a method is a minor change.

```
---

</p></details>

Showing results for e200d485c3e763085c924ce2763f507d27be1044
> Powered by [semver-audit-service](https://github.com/Workiva/semver-audit-service). Please report any problems by [filing an issue](https://github.com/Workiva/semver-audit-service/issues).
> Reported by the dart semver audit client 2.2.0
[Browse public API](https://semver-audit.wk-dev.wdesk.org/cleandart/react-dart/e200d485c3e763085c924ce2763f507d27be1044).

File: lib/react_client/react_interop.dart [hunk 5]
```diff
@@ -49,7 +49,7 @@ abstract class React {
   external static JsRef createRef();
   external static ReactClass forwardRef(Function(JsMap props, JsRef ref) wrapperFunction);
   external static ReactClass memo(
-    dynamic Function(JsMap props, [JsMap legacyContext]) wrapperFunction, [
+    dynamic wrapperFunction, [
```
Comment (greglittlefield-wf) [inline lib/react_client/react_interop.dart:52]: I had to loosen this to be able to pass in JS `ReactClass`es; this should not be considered a breaking change

File: lib/react.dart [hunk 6]
```diff
@@ -32,6 +32,11 @@ typedef Error PropValidator<TProps>(TProps props, PropValidatorInfo info);
 /// [props] is typed as [JsBackedMap] so that dart2js can make optimize props accesses.
 typedef DartFunctionComponent = dynamic Function(JsBackedMap props);
 
+/// The callback to a React forwardRef component. See [forwardRef2] for more details.
+///
+/// [props] is typed as [JsBackedMap] so that dart2js can make optimize props accesses.
```
Comment (smaifullerton-wk) [inline lib/react.dart:37]: #nit `make optimize` => `optimize`

File: test/factory/common_factory_tests.dart [hunk 7]
```diff
@@ -185,18 +216,79 @@ void domEventHandlerWrappingTests(ReactComponentFactoryProxy factory) {
...
+          propsFromDartRender = p;
+        },
+        dart.eventPropKey: (event) => events[dart] = event,
+      });
+
+      element = React.cloneElement(
+        element,
+        jsifyAndAllowInterop({
+          jsCloned.eventPropKey: (event) => events[jsCloned] = event,
+        }),
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply test/factory/common_factory_tests.dart:248]: This would be if a JS component clones a handler onto a ReactElement created by a Dart component.

For example, passing a Dart HOC into a MUI component that clones the HOC.

I'll add a comment!

[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply test/factory/common_factory_tests.dart:249]: Addressed in 70ebf83

File: test/forward_ref_test.dart [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import 'dart:js_util';
+
+import 'package:react/react.dart' as react;
+import 'package:react/react_client/react_interop.dart';
+import 'package:test/test.dart';
+
+import 'factory/common_factory_tests.dart';
+
+main() {
+  group('forwardRef2', () {
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply test/forward_ref_test.dart:47]: Reinstated in 8d52aa7

File: test/react_memo_test.dart [hunk 9]
```diff
@@ -27,7 +27,7 @@ main() {
             return prevProps['localCount'] == nextProps['localCount'];
           };
 
-    final MemoTest = react.memo(react.registerFunctionComponent((Map props) {
+    final MemoTest = react.memo2(react.registerFunctionComponent((Map props) {
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply test/react_memo_test.dart]: Reinstated in 8d52aa7

File: test/factory/dart_function_factory_test.dart [hunk 10]
```diff
@@ -45,21 +44,20 @@ main() {
...
-        isFunctionComponent: true,
+        MemoTest,
         dartComponentVersion: ReactDartComponentVersion.component2,
+        skipPropValuesTest: true,
       );
     });
   });
 
-  group('memo', () {
+  group('memo2', () {
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply test/factory/dart_function_factory_test.dart]: Moved to the react_memo_test.dart file in 8d52aa7

**Comment by greglittlefield-wf** [comment]
@aaronlademann-wf @smaifullerton-wk All feedback has been addressed! (including adding a test case for wrapping `forwardRef2` with `memo`)

File: test/react_memo_test.dart [hunk 11]
```diff
@@ -58,64 +111,62 @@ main() {
         reason: 'test setup sanity check');
   }
 
-  group('memo', () {
```
Comment (greglittlefield-wf) [inline test/react_memo_test.dart:61]: Best reviewed [with whitespace-agnostic diff](https://github.com/cleandart/react-dart/pull/275/files?w=1#diff-3c4e4fda8814c627565672f04c4999fdL61); this actually didn't change

**Review by aaronlademann-wf — APPROVED** [review]
+1

File: example/test/function_component_test.dart [hunk 12]
```diff
@@ -335,7 +335,7 @@ class FancyInputApi {
   FancyInputApi(this.focus);
 }
 
-final FancyInput = react.forwardRef((props, ref) {
+final FancyInput = react.forwardRef2((props, ref) {
   final inputRef = useRef();
 
   useImperativeHandle(
```
Comment (sydneyjodon-wk) [inline example/test/function_component_test.dart:341]: I'm getting a runtime error when serving this example. It looks like this `useImperativeHandle` is expecting a `Ref` object, but getting a `JsRef`?
Screen Shot 2020-09-08 at 11 15 32 AM

Comment (greglittlefield-wf) [reply example/test/function_component_test.dart:341]: Interesting... good catch, I'll look into this!

Comment (greglittlefield-wf) [reply example/test/function_component_test.dart:341]: Addressed in be896a7844a2f13e1a0a987266f9822b1fab965d!

File: test/hooks_test.dart [hunk 13]
```diff
@@ -642,97 +643,97 @@ main() {
     });
 
     group('useImperativeHandle -', () {
-      var mountNode = new DivElement();
```
Comment (greglittlefield-wf) [inline test/hooks_test.dart:645]: Best viewed with whitespace-agnostic diff; besides changing to `forwardRef2`, these were just indented

File: lib/src/react_client/dart_interop_statics.dart [hunk 14]
```diff
@@ -33,8 +33,9 @@ final ReactDartInteropStatics dartInteropStatics = (() {
           var ref = getProperty(jsThis.refs, name);
           if (ref == null) return null;
           if (ref is Element) return ref;
+          if (ref is ReactComponent) return ref.dartComponent ?? ref;
 
-          return (ref as ReactComponent).dartComponent ?? ref;
+          return ref;
```
Comment (greglittlefield-wf) [inline lib/src/react_client/dart_interop_statics.dart:38]: These changes were to fix an error when setting a string ref on a component using `useImperativeHandle` (because it would try to cast the customized ref to `ReactComponent`).

This an edge-case, but it was easier to fix the issue than to skip testing that case.

File: test/hooks_test.dart [hunk 15]
```diff
@@ -760,6 +761,19 @@ main() {
           expect(depsRef.current, 1, reason: 'current value updates because count.value is in dependency list');
         });
       });
+
+      group('works with all types of consumer refs -', () {
+        const customRefValue = {'Am I a ref?': true};
+        final factory = react.forwardRef2((props, ref) {
+          print('ref: $ref');
```
Comment (greglittlefield-wf) [inline test/hooks_test.dart:768]: LOG
```suggestion
```

**Comment by sydneyjodon-wk** [comment]
QAing now... I think this just needs to be formatted?

File: .travis.yml [hunk 16]
```diff
@@ -12,7 +12,10 @@ cache:
 
 before_script:
   - dartanalyzer .
-  - dartfmt --line-length=120 --dry-run --set-exit-if-changed .
+  - |
+    if [ "$TRAVIS_DART_VERSION" != "dev" ]; then
+      dartfmt --line-length=120 --dry-run --set-exit-if-changed .
```
Comment (greglittlefield-wf) [inline .travis.yml:17]: This fixes failures we were seeing in the Dart dev builds due to the dartfmt being different

**Comment by sydneyjodon-wk** [comment]
> - [x] Verify unit tests pass (CI passes)
> - [x] Smoke test examples
> - [x] Perform QA instructions in [over_react PR](https://github.com/Workiva/over_react/pull/620) 

QA+1 for every step except:

- [ ] Verify the function components in WSD that originally surfaced these issues still work (@aaronlademann-wf )

**Comment by aaronlademann-wf** [comment]
QA +1

- [x] Verify the function components in WSD that originally surfaced these issues still work
